# **Credit Card Fraud Detection Modelling.**

## **Load the train and test data sets.**

In [1]:
import os
import pandas as pd

DATASETS_PATH = os.getenv("DATASETS_PATH")
DATASETS_PATH = DATASETS_PATH + "/AnomalyDetection/CreditCardFraudDetection/"

initial_train_df = pd.read_csv(DATASETS_PATH + "fraudTrain.csv")

initial_test_df = pd.read_csv(DATASETS_PATH + "fraudTest.csv")

initial_train_df["trans_date_trans_time"] = pd.to_datetime(initial_train_df["trans_date_trans_time"])
initial_test_df["trans_date_trans_time"] = pd.to_datetime(initial_test_df["trans_date_trans_time"])

We are going to to build up features that describe "normal" behaviour of a customer. For that we are going to take temporal and spacial distributions and calculate Z-scores for them. By Z-Score a model is able to detect irregularities (deviation of a value from the normal distribution).

## **Create feature describing the credit card holder behaviour on the time axis by counting the minutes since midnight for each purchase.**

In [2]:
initial_train_df['minutes_of_day'] = initial_train_df['trans_date_trans_time'].dt.hour * 60 + initial_train_df['trans_date_trans_time'].dt.minute
initial_test_df['minutes_of_day'] = initial_test_df['trans_date_trans_time'].dt.hour * 60 + initial_test_df['trans_date_trans_time'].dt.minute

Calculate a map of average number of minutes of the day for a purchase and its deviation for each credit card holder, using the train dataset.

In [3]:
df_temporal_stats = initial_train_df.groupby('cc_num')['minutes_of_day'].agg(['mean', 'std']).reset_index()
df_temporal_stats.columns = ['cc_num', 'mean_minutes_of_day', 'std_minutes_of_day']

initial_train_df = initial_train_df.merge(df_temporal_stats, on='cc_num', how='left')
initial_test_df = initial_test_df.merge(df_temporal_stats, on='cc_num', how='left')

initial_train_df['temporal_z_score'] = (initial_train_df['minutes_of_day'] - initial_train_df['mean_minutes_of_day']) / initial_train_df['std_minutes_of_day']
initial_test_df['temporal_z_score'] = (initial_test_df['minutes_of_day'] - initial_test_df['mean_minutes_of_day']) / initial_test_df['std_minutes_of_day']


In [4]:
print(initial_train_df)

         Unnamed: 0 trans_date_trans_time               cc_num  \
0                 0   2019-01-01 00:00:18     2703186189652095   
1                 1   2019-01-01 00:00:44         630423337322   
2                 2   2019-01-01 00:00:51       38859492057661   
3                 3   2019-01-01 00:01:16     3534093764340240   
4                 4   2019-01-01 00:03:06      375534208663984   
...             ...                   ...                  ...   
1296670     1296670   2020-06-21 12:12:08       30263540414123   
1296671     1296671   2020-06-21 12:12:19     6011149206456997   
1296672     1296672   2020-06-21 12:12:32     3514865930894695   
1296673     1296673   2020-06-21 12:13:36     2720012583106919   
1296674     1296674   2020-06-21 12:13:37  4292902571056973207   

                                    merchant       category     amt  \
0                 fraud_Rippin, Kub and Mann       misc_net    4.97   
1            fraud_Heller, Gutmann and Zieme    grocery_pos  107.

## **Calculate the distribution of distances between merchants and a customer for all corresponding transactions.**

In [5]:
import numpy as np

def haversine(lat1, lon1, lat2, lon2):
    # convert degrees to radians
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])
    
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    
    a = np.sin(dlat / 2.0)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2.0)**2
    c = 2 * np.arcsin(np.sqrt(a))
    
    r = 6371  # Earth radius in kilometers
    return c * r

initial_train_df["distance_km"] = haversine(initial_train_df["lat"], initial_train_df["long"], initial_train_df["merch_lat"], initial_train_df["merch_long"])
initial_test_df["distance_km"] = haversine(initial_test_df["lat"], initial_test_df["long"], initial_test_df["merch_lat"], initial_test_df["merch_long"])

df_spacial_stats = initial_train_df.groupby('cc_num')['distance_km'].agg(['mean', 'std']).reset_index()
df_spacial_stats.columns = ['cc_num', 'mean_distance_km', 'std_distance_km']

initial_train_df = initial_train_df.merge(df_spacial_stats, on='cc_num', how='left')
initial_test_df = initial_test_df.merge(df_spacial_stats, on='cc_num', how='left')

initial_train_df['spacial_z_score'] = (initial_train_df['distance_km'] - initial_train_df['mean_distance_km']) / initial_train_df['std_distance_km']
initial_test_df['spacial_z_score'] = (initial_test_df['distance_km'] - initial_test_df['mean_distance_km']) / initial_test_df['std_distance_km']

print(initial_train_df)


         Unnamed: 0 trans_date_trans_time               cc_num  \
0                 0   2019-01-01 00:00:18     2703186189652095   
1                 1   2019-01-01 00:00:44         630423337322   
2                 2   2019-01-01 00:00:51       38859492057661   
3                 3   2019-01-01 00:01:16     3534093764340240   
4                 4   2019-01-01 00:03:06      375534208663984   
...             ...                   ...                  ...   
1296670     1296670   2020-06-21 12:12:08       30263540414123   
1296671     1296671   2020-06-21 12:12:19     6011149206456997   
1296672     1296672   2020-06-21 12:12:32     3514865930894695   
1296673     1296673   2020-06-21 12:13:36     2720012583106919   
1296674     1296674   2020-06-21 12:13:37  4292902571056973207   

                                    merchant       category     amt  \
0                 fraud_Rippin, Kub and Mann       misc_net    4.97   
1            fraud_Heller, Gutmann and Zieme    grocery_pos  107.

## **Recategorize the job descriptions for each customer using the ISCO 8 standard for jobs description.**

In [6]:
unique_jobs = initial_train_df["job"].fillna("").astype(str).unique()

In [7]:
isco_df = pd.read_excel(DATASETS_PATH + "ISCO-08 EN Structure and definitions.xlsx")

In [8]:
isco_df["Title EN"] = isco_df["Title EN"].fillna("")
isco_df["Definition"] = isco_df["Definition"].fillna("")
isco_df["Tasks include"] = isco_df["Tasks include"].fillna("")

print(isco_df)

     Level  ISCO 08 Code                                           Title EN  \
0        1             1                                           Managers   
1        2            11  Chief Executives, Senior Officials and Legisla...   
2        3           111                   Legislators and Senior Officials   
3        4          1111                                        Legislators   
4        4          1112                        Senior Government Officials   
..     ...           ...                                                ...   
614      3            21             Non-commissioned Armed Forces Officers   
615      4           210             Non-commissioned Armed Forces Officers   
616      2             3              Armed Forces Occupations, Other Ranks   
617      3            31              Armed Forces Occupations, Other Ranks   
618      4           310              Armed Forces Occupations, Other Ranks   

                                            Definit

Now we conduct the semantic similarity search between the job descriptions from the initial train data frame and standerdised job descriptions from the ISCO 8 document. The most similar descriptions from the standard will be chosen for further modelling.

In [9]:
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

def clean_text(series):
    return series.fillna("").astype(str)


initial_train_df["job"] = clean_text(initial_train_df["job"])
unique_jobs = initial_train_df["job"].unique().tolist()

isco_texts = (
    clean_text(isco_df["Title EN"]) + " " +
    clean_text(isco_df["Definition"]) + " " +
    clean_text(isco_df["Tasks include"])
).tolist()

model = SentenceTransformer("all-MiniLM-L6-v2")

job_embeddings = model.encode(unique_jobs, show_progress_bar=True)
isco_embeddings = model.encode(isco_texts, show_progress_bar=True)

similarity = cosine_similarity(job_embeddings, isco_embeddings)

best_match_idx = similarity.argmax(axis=1)
best_scores = similarity.max(axis=1)

mapping_df = pd.DataFrame({
    "job": unique_jobs,
    "isco_code": isco_df.iloc[best_match_idx]["ISCO 08 Code"].values.astype(int),
    "isco_title": isco_df.iloc[best_match_idx]["Title EN"].values,
    "similarity_score": best_scores
})

initial_train_df = initial_train_df.merge(mapping_df, on="job", how="left")

test_jobs = clean_text(initial_test_df["job"]).unique().tolist()

test_embeddings = model.encode(test_jobs, show_progress_bar=True)

similarity_test = cosine_similarity(test_embeddings, isco_embeddings)

best_idx_test = similarity_test.argmax(axis=1)

test_mapping = pd.DataFrame({
    "job": test_jobs,
    "isco_code": isco_df.iloc[best_idx_test]["ISCO 08 Code"].values,
    "isco_title": isco_df.iloc[best_idx_test]["Title EN"].values,
    "similarity_score": similarity_test.max(axis=1)
})

initial_test_df = initial_test_df.merge(test_mapping, on="job", how="left")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Batches:   0%|          | 0/16 [00:00<?, ?it/s]

Batches:   0%|          | 0/20 [00:00<?, ?it/s]

Batches:   0%|          | 0/15 [00:00<?, ?it/s]

In [10]:
print(initial_train_df[["job","isco_code","isco_title","similarity_score"]])
print()
print(initial_test_df[["job","isco_code","isco_title","similarity_score"]])

                                       job  isco_code  \
0                Psychologist, counselling       2634   
1        Special educational needs teacher       2352   
2              Nature conservation officer       2133   
3                          Patent attorney        261   
4           Dance movement psychotherapist       2653   
...                                    ...        ...   
1296670                       Geoscientist       2114   
1296671   Production assistant, television       2654   
1296672                    Naval architect       3152   
1296673              Volunteer coordinator       5312   
1296674           Therapist, horticultural       9214   

                                              isco_title  similarity_score  
0                                          Psychologists          0.516353  
1                                 Special Needs Teachers          0.714934  
2                 Environmental Protection Professionals          0.421878  
3      

## **Calculate job frequency feature for each category.**

Converting it to one hot encoding would explode dimensionality, that is why we choose frequency based categorization.

In [11]:
isco_freq = (
    initial_train_df["isco_code"]
    .value_counts()
    .to_dict()
)

initial_train_df["isco_code_freq"] = initial_train_df["isco_code"].map(isco_freq)

initial_test_df["isco_code_freq"] = initial_test_df["isco_code"].map(isco_freq)

global_mean_freq = initial_train_df["isco_code_freq"].mean()

initial_test_df["isco_code_freq"] = initial_test_df["isco_code_freq"].fillna(global_mean_freq)

print(initial_test_df)

        Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_care   29.84   Joanne   
2       frau

## **Transform the variable purchase category to a frequency map.**

In [12]:
category_map = (
    initial_train_df["category"]
    .value_counts()
    .to_dict()
)

initial_train_df["category_freq"] = initial_train_df["category"].map(category_map)
initial_test_df["category_freq"] = initial_test_df["category"].map(category_map).fillna(-1).astype(int)

In [13]:
print(initial_train_df[["category", "category_freq"]])
print()
print(initial_test_df[["category", "category_freq"]])
print()

              category  category_freq
0             misc_net          63287
1          grocery_pos         123638
2        entertainment          94014
3        gas_transport         131659
4             misc_pos          79655
...                ...            ...
1296670  entertainment          94014
1296671    food_dining          91461
1296672    food_dining          91461
1296673    food_dining          91461
1296674    food_dining          91461

[1296675 rows x 2 columns]

              category  category_freq
0        personal_care          90758
1        personal_care          90758
2       health_fitness          85879
3             misc_pos          79655
4               travel          40507
...                ...            ...
555714  health_fitness          85879
555715       kids_pets         113035
555716       kids_pets         113035
555717          travel          40507
555718   entertainment          94014

[555719 rows x 2 columns]



## **Transform variable zip to frequency map.**

In [14]:
zip_map = (
    initial_train_df["zip"]
    .value_counts()
    .to_dict()
)

initial_train_df["zip_freq"] = initial_train_df["zip"].map(zip_map)
initial_test_df["zip_freq"] = initial_test_df["zip"].map(zip_map).fillna(-1).astype(int)

In [15]:
print(initial_train_df[["zip", "zip_freq"]])
print()
print(initial_test_df[["zip", "zip_freq"]])
print()

           zip  zip_freq
0        28654      2028
1        99160      3030
2        83252       503
3        59632       493
4        24433      2017
...        ...       ...
1296670  84735      1513
1296671  21790       531
1296672  88325      2070
1296673  57756      2024
1296674  59871      1512

[1296675 rows x 2 columns]

          zip  zip_freq
0       29209      1561
1       84002      2089
2       11710      2577
3       32780      1526
4       49632      2034
...       ...       ...
555714  63453      1567
555715  77566      2556
555716  99323      2572
555717  83643      2038
555718  73034      2561

[555719 rows x 2 columns]



## **Transform the category gender to one hot encoding.**

In [16]:
# Train encoding
gender_ohe_train = pd.get_dummies(initial_train_df["gender"], prefix="gender", drop_first=True)

initial_train_df = pd.concat([initial_train_df, gender_ohe_train], axis=1)

# Test encoding (use TRAIN structure)
gender_ohe_test = pd.get_dummies(initial_test_df["gender"], prefix="gender")

# Align test to train columns
gender_ohe_test = gender_ohe_test.reindex(columns=gender_ohe_train.columns, fill_value=0)

initial_test_df = pd.concat([initial_test_df, gender_ohe_test], axis=1)

## **Create new variable 'age' of a credit card holder from the date of birth.**

In [17]:
from datetime import date

def compute_age(df):
    df["dob"] = pd.to_datetime(df["dob"], errors="coerce")
    
    today = pd.Timestamp(date.today())
    
    age = (
        today.year - df["dob"].dt.year
        - (
            (today.month < df["dob"].dt.month) |
            ((today.month == df["dob"].dt.month) & (today.day < df["dob"].dt.day))
        )
    )
    
    return age.astype("Int64")

initial_train_df["age"] = compute_age(initial_train_df)
initial_test_df["age"] = compute_age(initial_test_df)

In [18]:
print(initial_train_df[["dob","age"]])
print()
print(initial_test_df[["dob","age"]])

               dob  age
0       1988-03-09   38
1       1978-06-21   47
2       1962-01-19   64
3       1967-01-12   59
4       1986-03-28   40
...            ...  ...
1296670 1961-11-24   64
1296671 1979-12-11   46
1296672 1967-08-30   58
1296673 1980-08-18   45
1296674 1995-08-16   30

[1296675 rows x 2 columns]

              dob  age
0      1968-03-19   58
1      1990-01-17   36
2      1970-10-21   55
3      1987-07-25   38
4      1955-07-06   70
...           ...  ...
555714 1966-02-13   60
555715 1999-12-27   26
555716 1981-11-29   44
555717 1965-12-15   60
555718 1993-05-10   32

[555719 rows x 2 columns]


## **Calculate the purchase amount distribution for each person.**

In [19]:
initial_train_df["log_amt"] = np.log1p(initial_train_df["amt"])
initial_test_df["log_amt"] = np.log1p(initial_test_df["amt"])

df_amount_stats = initial_train_df.groupby("cc_num")["log_amt"].agg(["mean", "std"]).reset_index()
df_amount_stats.columns = ["cc_num", "mean_log_amt", "std_log_amt"]

initial_train_df = initial_train_df.merge(df_amount_stats, on="cc_num", how="left")
initial_test_df = initial_test_df.merge(df_amount_stats, on="cc_num", how="left")

initial_train_df["amt_z_score"] = (
    (initial_train_df["log_amt"] - initial_train_df["mean_log_amt"]) /
    initial_train_df["std_log_amt"]
)

initial_test_df["amt_z_score"] = (
    (initial_test_df["log_amt"] - initial_test_df["mean_log_amt"]) /
    initial_test_df["std_log_amt"]
)

In [20]:
print(initial_train_df)

         Unnamed: 0 trans_date_trans_time               cc_num  \
0                 0   2019-01-01 00:00:18     2703186189652095   
1                 1   2019-01-01 00:00:44         630423337322   
2                 2   2019-01-01 00:00:51       38859492057661   
3                 3   2019-01-01 00:01:16     3534093764340240   
4                 4   2019-01-01 00:03:06      375534208663984   
...             ...                   ...                  ...   
1296670     1296670   2020-06-21 12:12:08       30263540414123   
1296671     1296671   2020-06-21 12:12:19     6011149206456997   
1296672     1296672   2020-06-21 12:12:32     3514865930894695   
1296673     1296673   2020-06-21 12:13:36     2720012583106919   
1296674     1296674   2020-06-21 12:13:37  4292902571056973207   

                                    merchant       category     amt  \
0                 fraud_Rippin, Kub and Mann       misc_net    4.97   
1            fraud_Heller, Gutmann and Zieme    grocery_pos  107.

## **Create new data frames consisting of the generated features for training and tests.**

In [21]:
train_features_df = initial_train_df[["cc_num","gender_M","age","isco_code_freq","zip_freq","city_pop","category_freq","temporal_z_score","spacial_z_score","amt_z_score"]]

test_features_df = initial_test_df[["cc_num","gender_M","age","isco_code_freq","zip_freq","city_pop","category_freq","temporal_z_score","spacial_z_score","amt_z_score"]]

train_labels_df = initial_train_df["is_fraud"]

test_labels_df = initial_test_df["is_fraud"]

print(train_features_df)

                      cc_num  gender_M  age  isco_code_freq  zip_freq  \
0           2703186189652095     False   38           19356      2028   
1               630423337322     False   47           13268      3030   
2             38859492057661      True   64           11665       503   
3           3534093764340240      True   59            3553       493   
4            375534208663984      True   40            2036      2017   
...                      ...       ...  ...             ...       ...   
1296670       30263540414123      True   64           24132      1513   
1296671     6011149206456997      True   46           35043       531   
1296672     3514865930894695      True   58            8684      2070   
1296673     2720012583106919      True   45            2557      2024   
1296674  4292902571056973207      True   30            4622      1512   

         city_pop  category_freq  temporal_z_score  spacial_z_score  \
0            3495          63287         -2.231986  

## **Calculate anomaly score by Isolation Forest.**

In [22]:
from sklearn.ensemble import IsolationForest

train_features_df = train_features_df.drop(columns="cc_num")
test_features_df = test_features_df.drop(columns="cc_num")

model = IsolationForest(
    n_estimators=100,
    contamination=0.01,
    random_state=42
)

model.fit(train_features_df)

# anomaly score (higher = more normal in sklearn)
train_features_df["anomaly_score"] = model.decision_function(train_features_df)
test_features_df["anomaly_score"] = model.decision_function(test_features_df)

In [23]:
print(train_features_df)
print()
print(test_features_df)

         gender_M  age  isco_code_freq  zip_freq  city_pop  category_freq  \
0           False   38           19356      2028      3495          63287   
1           False   47           13268      3030       149         123638   
2            True   64           11665       503      4154          94014   
3            True   59            3553       493      1939         131659   
4            True   40            2036      2017        99          79655   
...           ...  ...             ...       ...       ...            ...   
1296670      True   64           24132      1513       258          94014   
1296671      True   46           35043       531       100          91461   
1296672      True   58            8684      2070       899          91461   
1296673      True   45            2557      2024      1126          91461   
1296674      True   30            4622      1512       218          91461   

         temporal_z_score  spacial_z_score  amt_z_score  anomaly_score  
0 

## **Conduct information gain analysis on generated features.**

In [24]:
from sklearn.feature_selection import mutual_info_classif

mi_train = mutual_info_classif(train_features_df, train_labels_df, discrete_features="auto")

mi_train_series = pd.Series(mi_train, index=train_features_df.columns).sort_values(ascending=False)

print(mi_train_series)

gender_M            0.127904
category_freq       0.027075
amt_z_score         0.013280
age                 0.005395
isco_code_freq      0.005381
zip_freq            0.005352
city_pop            0.004462
temporal_z_score    0.004091
anomaly_score       0.003386
spacial_z_score     0.000000
dtype: float64


As one can see the variable "spacial_z_score" does not reduce information disorder for the target of classification, or has zero variance on its own. Further this column will be excluded.

In [25]:
train_features_df = train_features_df.drop(columns="spacial_z_score")
test_features_df = test_features_df.drop(columns="spacial_z_score")

## **Compare three models Random Forests, XGBoost and LightGBM with manually tuned parameters in order to dismiss the worst performing ones.**

In [26]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import average_precision_score

model = RandomForestClassifier(
    n_estimators=450,
    verbose=1,
    max_depth=12,
    min_samples_split=10,
    min_samples_leaf=5,
    max_features="sqrt",
    class_weight="balanced",
    n_jobs=-1,
    random_state=42
)

print("Start fitting")
model.fit(train_features_df, train_labels_df)
print("Done")

y_prob = model.predict_proba(test_features_df)[:, 1]
test_pred_labels = (y_prob >= 0.5).astype(int)

y_true = test_labels_df

pr_auc = average_precision_score(y_true, y_prob)

print("Random Forest Performance:")
print("PR-AUC:", pr_auc)
print("Accuracy:", accuracy_score(y_true, test_pred_labels))
print("Precision:", precision_score(y_true, test_pred_labels))
print("Recall:", recall_score(y_true, test_pred_labels))
print("F1-score:", f1_score(y_true, test_pred_labels))

Start fitting


[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done  18 tasks      | elapsed:   18.5s
[Parallel(n_jobs=-1)]: Done 168 tasks      | elapsed:  1.8min
[Parallel(n_jobs=-1)]: Done 418 tasks      | elapsed:  4.4min
[Parallel(n_jobs=-1)]: Done 450 out of 450 | elapsed:  4.7min finished
[Parallel(n_jobs=16)]: Using backend ThreadingBackend with 16 concurrent workers.
[Parallel(n_jobs=16)]: Done  18 tasks      | elapsed:    0.0s


Done


[Parallel(n_jobs=16)]: Done 168 tasks      | elapsed:    1.0s
[Parallel(n_jobs=16)]: Done 418 tasks      | elapsed:    2.6s
[Parallel(n_jobs=16)]: Done 450 out of 450 | elapsed:    2.8s finished


Random Forest Performance:
PR-AUC: 0.8130798019830552
Accuracy: 0.9923900388505702
Precision: 0.32362897765741366
Recall: 0.8913752913752914
F1-score: 0.474854091642866


We keep the results of the random forest as a baseline and try to outperform it with more robust and advanced gradient boosted models.

The models are tuned manualy by iteratively changing parameters to the maximum of the PR-AUC one after another.

In [27]:
from xgboost import XGBClassifier

scale_pos_weight = (len(train_labels_df) - sum(train_labels_df)) / sum(train_labels_df)

model_xgb = XGBClassifier(
    n_estimators=200,
    max_depth=40,
    learning_rate=0.05,
    subsample=0.8,
    colsample_bytree=0.9,
    random_state=42,
    scale_pos_weight=scale_pos_weight,
    eval_metric="logloss"
)

model_xgb.fit(train_features_df, train_labels_df)

y_prob = model_xgb.predict_proba(test_features_df)[:, 1]
test_pred_labels = (y_prob >= 0.5).astype(int)

y_true = test_labels_df

pr_auc = average_precision_score(y_true, y_prob)
print("XGBoost Performance:")
print("PR-AUC:", pr_auc)
print("Accuracy:", accuracy_score(y_true, test_pred_labels))
print("Precision:", precision_score(y_true, test_pred_labels))
print("Recall:", recall_score(y_true, test_pred_labels))
print("F1-score:", f1_score(y_true, test_pred_labels))

XGBoost Performance:
PR-AUC: 0.8693169415604645
Accuracy: 0.9985082388761226
Precision: 0.8074766355140187
Recall: 0.8055944055944056
F1-score: 0.806534422403734


In [28]:
from lightgbm import LGBMClassifier

scale_pos_weight = (len(train_labels_df) - sum(train_labels_df)) / sum(train_labels_df)

model_lgbm = LGBMClassifier(
    n_estimators=900,
    max_depth=30,
    learning_rate=0.05,
    subsample=0.8,          
    colsample_bytree=0.7,   
    random_state=42,
    scale_pos_weight=scale_pos_weight
)

model_lgbm.fit(train_features_df, train_labels_df)

y_prob = model_lgbm.predict_proba(test_features_df)[:, 1]
test_pred_labels = (y_prob >= 0.5).astype(int)

y_true = test_labels_df

print("LighGBM Performance:")
print("PR-AUC:", average_precision_score(y_true, y_prob))
print("Accuracy:", accuracy_score(y_true, test_pred_labels))
print("Precision:", precision_score(y_true, test_pred_labels))
print("Recall:", recall_score(y_true, test_pred_labels))
print("F1-score:", f1_score(y_true, test_pred_labels))

[LightGBM] [Info] Number of positive: 7506, number of negative: 1289169
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.028628 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1566
[LightGBM] [Info] Number of data points in the train set: 1296675, number of used features: 9
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.005789 -> initscore=-5.146050
[LightGBM] [Info] Start training from score -5.146050
LighGBM Performance:
PR-AUC: 0.846913472220433
Accuracy: 0.9935327026788718
Precision: 0.3627580981246448
Recall: 0.8927738927738927
F1-score: 0.5158943965517241


## **Set best performing model.**

Comparing the baseline Random Forest, XGBoost and lightGBM, for which we tuned the hyperparameters manually aiming for maximum PR-AUC, one can conclude that XGBoost is the best performer here. So we choose the XGBoost model. 

In [29]:
best_model = model_xgb

y_prob = best_model.predict_proba(test_features_df)[:, 1]
test_pred_labels = (y_prob >= 0.5).astype(int)

y_true = test_labels_df

print("Best Performance:")
print("PR-AUC:", average_precision_score(y_true, y_prob))
print("Accuracy:", accuracy_score(y_true, test_pred_labels))
print("Precision:", precision_score(y_true, test_pred_labels))
print("Recall:", recall_score(y_true, test_pred_labels))
print("F1-score:", f1_score(y_true, test_pred_labels))



Best Performance:
PR-AUC: 0.8693169415604645
Accuracy: 0.9985082388761226
Precision: 0.8074766355140187
Recall: 0.8055944055944056
F1-score: 0.806534422403734


## **Find out how the probability threshold influences the fraud detection performance.**

In [30]:
thresholds = [0.3,0.55,0.7]

for threshold in thresholds:
    y_prob = best_model.predict_proba(test_features_df)[:, 1]
    test_pred_labels = (y_prob >= threshold).astype(int)
    
    y_true = test_labels_df
    
    print("Best Performance with threshold = "+ str(threshold) +" : ")
    print("PR-AUC:", average_precision_score(y_true, y_prob))
    print("Accuracy:", accuracy_score(y_true, test_pred_labels))
    print("Precision:", precision_score(y_true, test_pred_labels))
    print("Recall:", recall_score(y_true, test_pred_labels))
    print("F1-score:", f1_score(y_true, test_pred_labels))
    print()
    print()

Best Performance with threshold = 0.3 : 
PR-AUC: 0.8693169415604645
Accuracy: 0.9982491151103345
Precision: 0.7474662162162162
Recall: 0.8251748251748252
F1-score: 0.7844006204298692


Best Performance with threshold = 0.55 : 
PR-AUC: 0.8693169415604645
Accuracy: 0.998556824582208
Precision: 0.8211382113821138
Recall: 0.8004662004662004
F1-score: 0.8106704438149197


Best Performance with threshold = 0.7 : 
PR-AUC: 0.8693169415604645
Accuracy: 0.9986575949355699
Precision: 0.8538189175518462
Recall: 0.786946386946387
F1-score: 0.8190198932557011




**Multi-threshold strategy:**
* If predicted fraud probabililty is $>$ 0.70 - block transaction
* If it is between 0.30 and 0.70 - send to manual review
* If it is $<$ 0.30 - approve transaction

**This gives:**
* high precision where it matters
* high recall overall

## **Conclusions:**
* It was possible to create a well performing model for this simulated card fraud detection dataset. The PR-AUC $\approx 0.87$ with fraude rate of $\approx 0.6 \% $ for a synthetic credit card fraud dataset speaks for a **very strong model and features set**. It captures the underlying signals, but it cannot be graded as production model for fraud detection, because real world datasets are much more complicated and fraud behaviour adapts.
* The techniques used for this simulated datset can be though applied to realistic ones.